proposals/xxxx-super-call-enforcement.md

```diff
@@ -0,0 +1,200 @@
+}
+```
+
+An message can be (optionally) specified on the `@requiresSuper` attribute to provide any additional information, which will be shown with the error message:
+
+```swift
+class C1 {
+  @requiresSuper("Call super as the final step in your implementation")
+  func foo() {}
+}
+
+class C2: C1 {
+  override func foo() {} // error: method override is missing 'super.foo()' call: Call super as the final step in your implementation
+}
+```
+
+Any overrides of a method annotated with `@requiresSuper` will also implicitly inherit the attribute, to make sure that all the overrides in a subclass chain call back to the `super` method. This will be suppressed if the override is in a `final` class:
+
+```swift
+class C1 {
+  @requiresSuper
+  func foo() {}
+}
+
+class C2: C1 {
+  // Implicitly inherits '@requiresSuper' from 'C1.foo'
+  override func foo() {
+    super.foo()
+    // Do some other stuff
+  }
+}
+
+final class C3: C1 {
+  // Does not implicitly inherits '@requiresSuper' from 'C1.foo', since 'C3.foo' cannot be overridden
+  override func foo {
+    super.foo()
+    // Do some other stuff
+  }
+}
+
+class C4: C2 {
+  override func foo() {} // error: method override is missing 'super.foo()' call
+}
+```
+
+Finally, a `@requiresSuper` attribute will also be implicitly added for imported ObjC methods which have been annotated with `objc_requires_super`. This enables warnings (instead of errors) for any method with such an attribute, such as `UIViewController.viewDidLoad()`:
+
+```swift
+import UIKit
+
+class MyViewController1: UIViewController {
+  override func viewDidLoad() { } // warning: method override is missing 'super.viewDidLoad()' call
+}
+
+class MyViewController2: UIViewController {
+  override func viewDidLoad() { super.viewDidLoad() } // Okay
+}
+```
+
+## Source compatibility
+
+This is an additive change, so it does not break any existing Swift code. An imported ObjC declaration annotated with `objc_requires_super` will trigger a warning (instead of an error) if the overridden method skips the `super` call, which helps preserve source compatibility while giving users time to update their code.
+
+## Effect on ABI stability
+
+This change has no effect on the ABI.
+
+## Effect on API resilience
+
+Adding `@requiresSuper` to any existing API method can break code if an override of that method does not already call `super` in its body. Removing the attribute does not have any effect.
+
+## Alternatives considered
+
+- Do not diagnose missing 'super' calls.
+- Diagnose missing 'super' calls but also enforce its placement: The ordering of a 'super' call (first, middle, last, etc) is not syntactical requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicitly, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times its called.
```

Comment:
```suggestion
- Diagnose missing 'super' calls but also enforce their placement: The ordering of a 'super' call (first, middle, last, etc) is not a syntactic requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicitly, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times its called.
```